### PR TITLE
Add CLI arg: --with-pattern

### LINKED ISSUE #1979: Make --print-queue output match the required input of --enqueue & control file
I may have mentioned before that I have a script to manage quodlibet's queue with Dropbox for me. Well, I've run into major trouble with that script. I've been trying to modify it to support queues that are longer than bash's maximum amount of arguments. Your first thought should be: why should the max args even be an issue when I can pipe it into something? The answer is, I can't, because --print-queue gives me crap that does NOT match the input syntax of commands like enqueue, like file:// and HTML character codes, and to fix that, I have to do this:
`python2 -c "import sys, urllib as ul; print ul.unquote_plus(sys.argv[1])" "$(quodlibet --print-queue | sed 's|file://||g')"`
Which, now that I've modified it to support long queues, has turned into:
`quodlibet --print-queue | xargs -P 6 -I{} -d '\n' python2 -c "import sys, urllib as ul; print ul.unquote_plus(sys.argv[1])" "{}" | sed 's|file://||g'`
Long story short, it refused to work without the -I{} option, which slows things way down and makes this way too unwieldy, which the -P 6 option does not do a good enough job of fixing.

If --print-queue would just have proper output--namely by feeding it through that python code I'm using IN the program itself before outputting--I could just do `quodlibet --print-queue` with no problem.

### LINKED ISSUE #3080: command enhancements: print-query, print-playlist
Currently the `--print-playlist` and `--print-query` commands only return the songs' paths & filenames.  The `--print-playing` command however allows a tag pattern argument and, if present, returns the matching string. Could the `--print-playlist` and `--print-query` commands be enhanced to return strings matching an optional tag pattern?

eg 
`quodlibet --print-query=performer:guitar='John McLaughlin'  '<~album~title>'`

also, a blank query with a tag pattern should also print all the existing tag values:
`quodlibet --print-query= '<genre>'`

### PR DESCRIPTION
**Pull request by nemethf**
Closes #1979
Closes #3080

Check-list
----------

 * [x] There is a linked issue discussing the motivations for this feature or bugfix
 * [x] Unit tests have been added where possible
 * [x] I've added / updated documentation for any user-facing features.
 * [x] Performance seems to be comparable or better than current `master`

What this change is adding / fixing
-----------------------------------

This is the second iteration of #4143.  It addresses the reviews of the original pull requests.   Performance-wise it should be the same.  I've also extended the documentation.   Caveat: I wrote the shell completion parts in a cargo-cult style and haven't tested them.  Also, I don't really understand the senf library, but followed existing usage.

### CODE REVIEW

**Comment by nemethf** [comment]
Once again, I have no idea what msys2 test failure means.  Hopefully, it's the tests and not the pull request.

**Comment by afontenot** [comment]
This looks good to me, with one exception. I don't really love the `extract_args_from_string` approach. Most of the command running code is already built to support multiple arguments. I think it would be nicer to try to get Quod Libet to pass arbitrary data. AFAIK the only reason this hasn't been done is that until now we haven't needed to pass multiple arguments for anything.

Probably the best way to do this would be to serialize the command and arguments as JSON, e.g.

    {"command": "some_command", "args": ["value 1", "value 2"], "kwargs": {"named_arg1": "value 3"}

On the other hand, I'm not against merging this as is and going back to improve it later.

> Once again, I have no idea what msys2 test failure means. Hopefully, it's the tests and not the pull request.

MSYS is Windows. It's not clear to me what the cause of the failure is.

**Comment by nemethf** [comment]
> This looks good to me, with one exception. I don't really love the `extract_args_from_string` approach.  [...] 
> Probably the best way to do this would be to serialize the command and arguments as JSON, [...]

I force-pushed a new version that implements json encoding for the arguments of the print-query command and I opened a new issue (#4170) to discuss how the general argument handling could be changed.

**Comment by afontenot** [comment]
I think this looks quite nice! Solves the problem while leaving room for future progress, e.g. by switching all the communication to serialized JSON or going with one of your ideas discussed in #4170.

One question - do you need the backwards compatibility stuff? Is it necessary that an old version of Quod Libet be able to send arguments to a newer version? I'm not saying it isn't, I really don't know what the requirements are for this...

**Comment by nemethf** [comment]
> do you need the backwards compatibility stuff? 

I don't.  I just wasn't sure whether it is necessary to provide it.

File: tests/test_commands.py [hunk 1]
```diff
@@ -109,3 +109,24 @@ def test_enqueue_files(self):
         self.__send("enqueue-files "
                     "one,two\\, please,slash\\\\.mp3,four")
         self.assertEquals(app.window.playlist.q.get(), songs)
+
+    def test_with_pattern(self):
```
Comment (declension) [inline tests/test_commands.py:113]: I think it would be good to test the case of some valid JSON that's also a query

`"foo"`
(or even `1.0` etc)

Comment (nemethf) [reply tests/test_commands.py:113]: You're right.  This actually revealed a bug in the PR.

File: quodlibet/commands.py [hunk 2]
```diff
@@ -525,14 +535,22 @@ def _refresh(app):
...
     """Queries library, dumping filenames of matches to stdout
     See Issue 716
     """
 
-    query = arg2text(query)
+    try:
+        args = json.loads(arg2text(json_encoded_args))
+        query = args['query']
+        fstring = args['pattern']
+    except (json.decoder.JSONDecodeError, IndexError):
```
[the remark this answers is not in the document]
Comment (nemethf) [reply quodlibet/commands.py]: Yeah, you're right

**Comment by nemethf** [comment]
@declension, thank your for the review.  I pushed a new version addressing your concerns.

File: tests/test_commands.py [hunk 3]
```diff
@@ -109,3 +109,46 @@ def test_enqueue_files(self):
...
+        songs = [AudioFile({"~filename": fn, "title": fn.upper()})
+                 for fn in ["one", "two, please", "slash\\.mp3", "4.0-four"]]
+        app.library.add(songs)
+
+        self.assertFalse(app.window.playlist.q.get())
+        self.__send("enqueue-files "
+                    "one,two\\, please,slash\\\\.mp3,4.0-four")
+        # Old syntax
+        assert self.__send("print-query two ") == "two, please\n"
+        # Old syntax that is also a valid json
```
Comment (declension) [inline tests/test_commands.py:123]: This is all much better :+1:

Could you split the assertions up into tests? Motivation: it's way easier to debug when a test called `test_old_syntax_that_is_valid_json()` starts failing, than "line 123 of _test_everything()`" etc

The comments pretty much become the test name, and anything shared is a fixture (ideally) or here perhaps just in `setUp()`

Comment (nemethf) [reply tests/test_commands.py]: I refactored these tests into a new class as requested.  (I think the resulting code is a bit complex.)  Thanks.

Comment (declension) [reply tests/test_commands.py]: Great, thanks!

**Comment by declension** [comment]
Thanks @nemethf

**Pull request merged**